reverse_image_search_module/test/test_multicrop.py

```diff
@@ -0,0 +1,149 @@
+import os
+import numpy as np
+import random
+from PIL import Image
+from ..search_image import find_index_from_image, find_file_name, find_repeating_index
+import logging
+import matplotlib.pyplot as plt
+from datetime import datetime
+
+
+logging.basicConfig(level=logging.INFO)
+mylogger = logging.getLogger()
+
+
+DATA_IMAGES_PATH = 'data/img/full/'
+K_CROPPING = 10
+N_SEARCH = 10
+IMAGE_MAX_NUMBER = 5000
+
+
+def random_cropping(image_width, image_height, min_size, K):
+    for _ in range(K):
+        x_ini = random.randint(0, image_width - min_size)
+        y_ini = random.randint(0, image_height - min_size)
+
+        x_end = random.randint(x_ini + min_size, image_width)
+        y_end = random.randint(y_ini + min_size, image_height)
+
+        yield x_ini, y_ini, x_end, y_end
+
+
+def get_image_matching_many(image_names, many=False):
+    positive = 0
+
+    for _im in image_names:
+        try:
+            image = Image.open(os.path.join(DATA_IMAGES_PATH, _im))
+            # to check the size and truncation
+            _ = image.resize((255, 255))
+        except Exception as ex:
+            mylogger.info(f'image {_im} load error')
+            mylogger.info(str(ex))
+            continue
+        
+        
+        idxs, dists = [], []
+
+        for x_ini, y_ini, x_end, y_end in random_cropping(255,255, 200, K_CROPPING):
+            image_croped = image.crop((x_ini, y_ini, x_end, y_end))
+
+            idx, dist = find_index_from_image(image_croped, n=N_SEARCH)
+
+            idx, dist = idx[0], dist[0]
+
+            # if many:
+            #     idx, dist = find_repeating_index(idx, dist)
+            # else:
+            #     idx, dist = idx[0], dist[0]
+            
+            idxs.append(idx)
+            dists.append(dist)
+        
+        selected_idx, selected_dist = max(zip(idxs, dists), key=lambda x: x[1])
+
+        file_name = find_file_name(selected_idx)
+
+        if file_name == _im:
+            positive += 1
```

Comment:
This may be a wrong look at this case, but logic like this (calculating the metric) is usually found one level higher. If this function could just find the closest and return the list of closest, you can then perform calculation of the metric. Please ignore if you feel this is nitpicking :)

Reply:
The other solution: rename the function. “Get image” in the name means you’re going to have some images maybe? But not the number which is a metric value. This should remove ambiguity for me.

Reply:
Upon looking at the next function, I think renaming is better since metric calculation is involved process and it will be unnecessary complicated to return all images, thresholds, etc. The only pro in that case is if you want to make these functions more flexible: like you want to score different approaches (1:1, m:1, etc) against different ways of calculating the match, I.e. the image you find (just max cosine, harmonic mean, etc).

Reply:
Resolved in commit 91e1a92 :)